src/cmds/qsub.c

```diff
@@ -4703,524 +4864,468 @@ handle_attribute_errors(struct ecl_attribute_errors *err_list,
-	 * the new child process. The foreground process waits on the event
-	 * object, till the background process is up, sets up the named pipe,
-	 * and signals this event to tell the foreground process to continue
-	 * and  try to connect to it via this new named pipe.
-	 */
-	{
-		HANDLE hFile;
-		SECURITY_ATTRIBUTES sa;
-		STARTUPINFO si = {sizeof(si)};
-		PROCESS_INFORMATION pi;
-		char cmd_line[2*MAXPATHLEN + 1];
-		int created = 0;
-		HANDLE hEvent;
+/**
+ * @brief
+ *	Helper function to duplicate the passed attrl structure.
+ *
+ * @param[in]	attrib - The list of attributes to copy
+ *
+ * @return	struct attrl * - the duplicated 'attrib' (malloced).
+ * @retval	non-NULL - success.
+ * @retval	NULL - failure.
+ *
+ */
+static struct attrl *
+dup_attrl(struct attrl *attrib)
+{
+	struct attrl *attr = NULL;
+	struct attrl *attr_new = NULL;
 
-again:
-		hFile = CreateFile(fl, GENERIC_READ | GENERIC_WRITE, 0, NULL,
-			OPEN_EXISTING, 0, NULL);
-		if (hFile == INVALID_HANDLE_VALUE) {
-			if (created == 0) {
-				sa.nLength = sizeof(sa);
-				sa.lpSecurityDescriptor = NULL;
-				sa.bInheritHandle = TRUE;
-
-				/* now create a named event to wait on later */
-				hEvent = CreateEvent(
-					&sa, /* default security attribute */
-					TRUE, /* manual-reset event */
-					FALSE, /* initial state = signaled */
-					NULL); /* unnamed event object */
-				if (hEvent == NULL)
-					goto regular_submit;
+	attr = attrib;
+	while (attr) {
+		if (attr->resource != NULL) {
+			/* strings have null character also in buf */
+			set_attr_resc(&attr_new, attr->name,
+				attr->resource,
+				attr->value);
+		} else {
+			set_attr(&attr_new, attr->name, attr->value);
+		}
 
-				/* launch new qsub process, connect 2 server */
-				sa.bInheritHandle = FALSE;
-				sprintf(cmd_line, "%s --daemon %s %d %s",
-					qsub_exe, fl,
-					hEvent, server_out);
-				if (!CreateProcess(NULL, cmd_line, &sa, &sa,
-					TRUE, CREATE_NO_WINDOW, NULL,
-					NULL, &si, &pi)) {
-					CloseHandle(hEvent);
-					goto regular_submit;
-				}
+		attr=attr->next;
+	}
 
-				/* now wait for single object */
-				/* foreground process wait a max 10 seconds */
-				rc = WaitForSingleObject(hEvent, 10 * 1000);
-				CloseHandle(hEvent);
+	return attr_new;
+}
 
-				if (rc != WAIT_OBJECT_0)
-					goto regular_submit; /* timeout */
+/**
+ * @brief
+ *	Helper function to get the pbs conf file path, and
+ *	convert it to a string, later added to the filename
+ *	(pipe or unix domain socket filename) that is used
+ *	for communications between the front-end and
+ *	background qsub processes.
+ *	The path to the pbs conf file is converted to a string
+ *	by replacing the slashes with underscrore char.
+ *	If PBS_CONF_FILE is not set, then an empty string
+ *	is returned.
+ *
+ * @return - The string representing path to the pbs conf file
+ *
+ */
+static char *
+get_conf_path()
+{
+	char *cnf = getenv("PBS_CONF_FILE");
+	char *dup_cnf_path = "";
+	char *p;
```

Comment:
Since using `realloc()` with `dup_cnf_path` would require copying the string returned by `strdup()` and then immediately freeing it, I instead used a static pointer that points directly to the string returned by `strdup()`. Each time `get_conf_path()` is called, it first frees the memory allocated by `strdup()` during the previous invocation of `get_conf_path()`, and then assigns `dup_cnf_path` to the new string returned by `strdup()`. 



This should provide the same functionality while using less memory and avoiding unnecessary copying.